// appium-remote-debugger/build/lib/remote-debugger-real-device.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RemoteDebuggerRealDevice = void 0;
const remote_debugger_1 = require("./remote-debugger");
const rpc_1 = require("./rpc");
class RemoteDebuggerRealDevice extends remote_debugger_1.RemoteDebugger {
    _udid;
    constructor(opts) {
        super(opts);
        this._udid = opts.udid;
        this._skippedApps = ['lockdownd'];
    }
    initRpcClient() {
        this._rpcClient = new rpc_1.RpcClientRealDevice({
            bundleId: this._bundleId,
            platformVersion: this._platformVersion,
            isSafari: this._isSafari,
            logAllCommunication: this._logAllCommunication,
            logAllCommunicationHexDump: this._logAllCommunicationHexDump,
            socketChunkSize: this._socketChunkSize,
            webInspectorMaxFrameLength: this._webInspectorMaxFrameLength,
            udid: this._udid,
            pageLoadTimeoutMs: this._pageLoadMs,
        });
    }
}
exports.RemoteDebuggerRealDevice = RemoteDebuggerRealDevice;
exports.default = RemoteDebuggerRealDevice;
